py/server/deephaven/experimental/s3.py

```diff
@@ -118,12 +145,15 @@ def __init__(self,
                     (access_key_id is None and secret_access_key is not None)):
                 raise DHError("Either both access_key_id and secret_access_key must be provided or neither")
 
+            # Only one of the following credentials can be set, otherwise the builder will throw an exception
             if access_key_id is not None:
-                if anonymous_access:
-                    raise DHError("Only one set of credentials may be used, requested both key and anonymous")
```

Comment:
Should this exception have been removed?  Is it acceptable to have an access key and anonymous?  Does the check happen in java? Is the resulting exception useful in Python?

Reply:
- No, we cannot specify access key and anonymous. I have tried to explain that in the comments. Let me know if its more clear.
- Yes, the check happens in the java code, I have added a comment in line 152 to explain that.
- The error would look like this:
```
RuntimeError: java.lang.IllegalStateException: Builder of S3Instructions is strict, attribute is already set: credentials
```
Is that okay?

Reply:
I don't think this is a clear exception for a user.  How is the user supposed to know where two sets of credentials come from?

Reply:
Also, no python argument is named `credentials`

Reply:
I have updated the logic, the error would now look like
```
Value: Failed to build S3 instructions : deephaven.dherror.DHError: Only one set of credentials can be set, but found anonymous_access and use_default_credentials : Only one set of credentials can be set, but found anonymous_access and use_default_credentials
```

Let me know if this is more clear.

Reply:
Another way to do this would be to add an enum for credentials, where user can specify they want to use the anonymous or profile, etc. Just that anonymous is also a boolean right now. And if I remove that, it will become a breaking change. Do you have any thoughts on this?

Reply:
enum may be the cleanest.  @jmao-denver Thoughts?

Reply:
Hi @chipkent, @jmao-denver and I had a quick discussion about the above and Jianfeng suggested to make a wrapper for the Credentials Java interface, to make the entire thing similar to Java. For example,

```
class Credentials(JObjectWrapper):

    j_object_type = _JCredentials or type(None)

    def __init__(self, _j_object: jpy.JType):
        self._j_object = _j_object

    @classmethod
    def basic(cls, access_key_id: str, secret_access_key: str) -> Credentials:
        return cls(_JCredentials.basic(access_key_id, secret_access_key))

    @classmethod
    def anonymous(cls) -> Credentials:
        return cls(_JCredentials.anonymous())

    @classmethod
    def profile(cls) -> Credentials:
        return cls(_JCredentials.profile())
```

I also like this design. We can do a breaking change to make things very clear so user can only set a Credentials object and move all the documentation to this new class.

Does this sound reasonable to you?

Reply:
Made this change, let me know what you both think.

Reply:
I'm not against this change.  It looks like it might end up cleaner.  Since this is in experimental, I'm not concerned about breaking users.  Once we move it out, I'll be concerned.  Experimental is so that we can iterate to get something good.